tcmur_cmd_handler.c

```diff
@@ -57,12 +57,15 @@ void tcmur_command_complete(struct tcmu_device *dev, struct tcmulib_cmd *cmd,
 static void aio_command_finish(struct tcmu_device *dev, struct tcmulib_cmd *cmd,
 			       int rc)
 {
-	int wakeup;
+	struct tcmur_device *rdev = tcmu_get_daemon_dev_private(dev);
+	int wake_up;
 
-	track_aio_request_finish(tcmu_get_daemon_dev_private(dev), &wakeup);
+	track_aio_request_finish(rdev, &wake_up);
 	tcmur_command_complete(dev, cmd, rc);
-	if (wakeup)
+	while (wake_up) {
 		tcmulib_processing_complete(dev);
+		track_aio_wakeup_finish(rdev, &wake_up);
+	}
 }
```

Comment:
I think we could hit a race:

1. pending_wakeups = 0. thread1 calls track_aio_request_finish, pending_wakeups is incremented to 1, and wake_up=1 is returned. It does it's call to tcmur_command_complete and then falls into the while loop and does tcmulib_processing_complete.

2. thread2 then calls track_aio_request_finish and increments pending_wakeups=2. wake_up=0 is returned so it is relying on thread1 to call tcmulib_processing_complete.

3. thread1 then calls track_aio_wakeup_finish. It sees aio_track->pending_wakeups > 1 and sets pending_wakeups=1and returns wake_up=1.

4. thread1 is super fast and we loop again and makes the tcmulib_processing_complete call and track_aio_wakeup_finish which should have completed the processing for the cmd for thread2. pending_wakeups=1 so it is now set to 0 and wake_up=0 is returned. thread1 is now done. It returns.

5. thread2 finally calls tcmur_command_complete (or it could have been in the middle of calling it but yet done updating the ring buffer). wake_up was 0 for it, so it does not go into the loop above and tcmulib_processing_complete is never called for it.

If the race analysis is correct I think we just need to do something like the attached where we hold the device completion lock while updating the aio track calls as well as doing the tcmulib_command_complete. aio_command_finish it just a little funky in that it straddles both the aio code and tcmur cmd handler code, so my change here is a little gross.

[Jasons392-plus-my-extra-dumb-locking.txt](https://github.com/open-iscsi/tcmu-runner/files/1878209/Jasons392-plus-my-extra-dumb-locking.txt)

Reply:
Couldn't the call to ```tcmur_command_complete``` and ```track_aio_request_finish``` be swapped?

Reply:
> Couldn't the call to tcmur_command_complete and track_aio_request_finish be swapped?

Lol, yeah that is much better! My eyes were probably going crosseyed :)

Reply:
OK -- swapped the two function calls.